### PR TITLE
Add API documentation for plugable secret backends

### PR DESCRIPTION
**Pull request by thaJeztah**
Documents the API changes introduced in

0304c98d85404fe75a1b4a35d3c111931e062f41 (https://github.com/moby/moby/pull/34157) and 08f7cf05268782a0dd8e4c41a4cc65fdf78d09f2 (https://github.com/moby/moby/pull/34123)

**- What I did**

**- How I did it**

**- How to verify it**

run `make swagger-docs` and visit http://localhost:9000 to view the docs

ping @liron-l  @cpuguy83 @vdemeester  PTAL

### CODE REVIEW

File: api/swagger.yaml [hunk 1]
```diff
@@ -2790,6 +2790,24 @@ definitions:
           type: "array"
           items:
             $ref: "#/definitions/Mount"
+
+  Driver:
+    description: "Driver represents a driver (network, logging, secrets)."
+    type: "object"
+    required: [Name, Data]
```
Comment (tophj-ibm) [inline api/swagger.yaml]: janky failure looks related. 

``` 
09:58:42 - The swagger spec at "api/swagger.yaml" is invalid against swagger specification 2.0. see errors :
09:58:42 - "Data" is present in required but not defined as property in definition "Driver"
09:58:42 - /secrets/{id}.Spec.Driver.Data in body is required
```

Q: I haven't looked at the swagger.yaml before, but I'm guessing Data is supposed to Options here, which is based off https://github.com/moby/moby/blob/0304c98d85404fe75a1b4a35d3c111931e062f41/api/types/swarm/common.go#L23 right?

Comment (thaJeztah) [inline api/swagger.yaml]: oh, hm, looks like I may have copy/pasted the `Data` yes thanks, I'll check

Comment (thaJeztah) [inline api/swagger.yaml]: Don't think `Data` / `Options` is required here, so I'll remove

**Comment by thaJeztah** [comment]
ping @liron-l @cpuguy83 @vdemeester PTAL

File: api/swagger.yaml [hunk 2]
```diff
@@ -2790,6 +2790,24 @@ definitions:
...
+  Driver:
+    description: "Driver represents a driver (network, logging, secrets)."
+    type: "object"
+    required: [Name]
+    properties:
+      Name:
+        type: "string"
+        x-nullable: false
+        example: "some-driver"
+      Options:
```
Comment (liron-l) [inline api/swagger.yaml]: Consider 
`description: "key/value map of driver specific options."` (same as `DriverConfig.Options)`

File: docs/api/version-history.md [hunk 3]
```diff
@@ -19,6 +19,14 @@ keywords: "API, Docker, rcli, REST, documentation"
 
 * `DELETE /secrets/(name)` now returns status code 404 instead of 500 when the secret does not exist.
 * `POST /secrets/create` now returns status code 409 instead of 500 when creating an already existing secret.
+* `POST /secrets/create` now accepts a `Driver` struct, allowing the
+  `Name` and diver-specific `Options` to be passed to store a secrets
```
Comment (liron-l) [inline docs/api/version-history.md]: diver-specific -> driver-specific (here and below)

Comment (thaJeztah) [inline docs/api/version-history.md]: Good catch :)

**Comment by liron-l** [comment]
LGTM @thaJeztah, left some minor comments inline.

**Comment by thaJeztah** [comment]
@liron-l thanks, I updated.

Question: do you know;

- What are the name restrictions on Secrets drivers? (e.g., only `a-zA-Z0-9_.-`)
- Is there a maximum length?
- Are Secrets driver names case-sensitive?
- If none of the above; can we add such validation? (We've been bitten by people relying on no restrictions being there, and once we wanted to add validation, it was a breaking change)

TL;DR I'd like to document what are acceptable values for the `Name` field :)

**Comment by liron-l** [comment]
Thanks @thaJeztah.
Technically, the restrictions should be on the secrets plugin name, since the driver name is only a reference to the name of the plugin. Also, we should enforce that the specified secret's plugin exist as part of the `docker secret create` flow (and then we don't need additional name validation).

Finally, if we do add such restrictions on the name, I know that plguins names need to have versioning support (`:`). 
@cpuguy83 does this makes sense?

**Comment by thaJeztah** [comment]
Ah, right, yes, these are managed plugins (installed through `docker plugin install`)? (sorry, I may be slow, ha!) in that case, we should probably be covered here (can add some information in future)

**Comment by liron-l** [comment]
Yes, @thaJeztah, secret plugins are managed plugins. We will add dedicated documentation and examples after all relevant commits are merged.
docker/cli#366
docker/swarmkit#2326

File: api/swagger.yaml [hunk 4]
```diff
@@ -2790,6 +2790,26 @@ definitions:
...
+        description: "Key/value map of driver-specific options."
+        type: "object"
+        x-nullable: false
+        additionalProperties:
+          type: "string"
+          example:
+            "value for driver-specific option"
```
Comment (dnephin) [inline api/swagger.yaml]: This formatting seems wrong. I think there's a key missing here on the newline?

File: api/swagger.yaml [hunk 5]
```diff
@@ -2802,10 +2822,19 @@ definitions:
         additionalProperties:
           type: "string"
       Data:
-        description: "Base64-url-safe-encoded secret data"
+        description: |
+          Base64-url-safe-encoded ([RFC 4648](https://tools.ietf.org/html/rfc4648#section-3.2))
+          data to store as secret.
+
+          This field is only used to _create_ a secret, and is not returned by
+          other endpoints.
```
Comment (dnephin) [inline api/swagger.yaml]: Hmm, we should probably have two different types then. One without this field. Since this already exited we can wait for another PR

File: api/swagger.yaml (hunk 4 above)
Comment (thaJeztah) [inline api/swagger.yaml]: oh, this is odd, yes, definitely should be fixed 👍

File: api/swagger.yaml (hunk 5 above)
Comment (thaJeztah) [inline api/swagger.yaml]: Yes; while working on these changes in the Swagger, I found other types that are reused and don't make sense

I think this is an "omitempty" (so in practice it's not there), but yes, probably should have a separate type

File: api/swagger.yaml [hunk 6]
```diff
@@ -2833,24 +2854,38 @@ definitions:
         type: "object"
         additionalProperties:
           type: "string"
+        example:
+          com.example.some-label: "some-value"
+          com.example.some-other-label: "some-other-value"
       Data:
-        description: "Base64-url-safe-encoded secret data"
-        type: "array"
```
Comment (thaJeztah) [inline api/swagger.yaml]: FWIW I changed this from an array to just string; I guess this was because in go it's a `[]byte`

screen shot 2017-07-27 at 23 53 47

**Comment by thaJeztah** [comment]
@dnephin updated PTAL

**Review by dnephin — APPROVED** [review]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐸
